src/analysis/normed_space/normed_group_hom_completion.lean

```diff
@@ -0,0 +1,259 @@
+by { ext x, exact normed_group_hom.completion_def f x }
+
+@[simp]
+lemma normed_group_hom.completion_coe (f : normed_group_hom G H) (g : G) : f.completion g = f g :=
+completion.map_coe f.uniform_continuous _
+
+/-- Completion of normed group homs as a normed group hom. -/
+def normed_group_hom_completion_hom :
+  normed_group_hom G H →+ normed_group_hom (completion G) (completion H) :=
+{ to_fun := normed_group_hom.completion,
+  map_zero' := begin
+    apply to_add_monoid_hom_injective,
+    exact add_monoid_hom.completion_zero
+  end,
+  map_add' := λ f g, begin
+    apply to_add_monoid_hom_injective,
+    exact f.to_add_monoid_hom.completion_add g.to_add_monoid_hom f.continuous g.continuous,
+  end }
+
+@[simp]
+lemma normed_group_hom.completion_id :
+  (normed_group_hom.id G).completion = normed_group_hom.id (completion G) :=
+begin
+  ext x,
+  rw [normed_group_hom.completion_def, normed_group_hom.coe_id, completion.map_id],
+  refl
+end
+
+lemma normed_group_hom.completion_comp (f : normed_group_hom G H) (g : normed_group_hom H K) :
+  g.completion.comp f.completion = (g.comp f).completion :=
+begin
+  ext x,
+  rw [normed_group_hom.coe_comp, normed_group_hom.completion_def,
+    normed_group_hom.completion_coe_to_fun, normed_group_hom.completion_coe_to_fun,
+    completion.map_comp (normed_group_hom.uniform_continuous _)
+    (normed_group_hom.uniform_continuous _)],
+  refl
+end
+
+lemma normed_group_hom.completion_neg (f : normed_group_hom G H) :
+  (-f).completion = -f.completion :=
+normed_group_hom_completion_hom.map_neg f
+
+lemma normed_group_hom.completion_add (f g : normed_group_hom G H) :
+  (f + g).completion = f.completion + g.completion :=
+normed_group_hom_completion_hom.map_add f g
+
+lemma normed_group_hom.completion_sub (f g : normed_group_hom G H) :
+  (f - g).completion = f.completion - g.completion :=
+normed_group_hom_completion_hom.map_sub f g
+
+/-- The map from a normed group to its completion, as a normed group hom. -/
+def normed_group.to_compl : normed_group_hom G (completion G) :=
+{ to_fun := coe,
+  map_add' := completion.to_compl.map_add,
+  bound' := ⟨1, by simp [le_refl]⟩ }
+
+open normed_group
+
+lemma normed_group.norm_to_compl (x : G) : ∥to_compl x∥ = ∥x∥ :=
+completion.norm_coe x
+
+@[simp]
+lemma normed_group_hom.zero_completion : (0 : normed_group_hom G H).completion = 0 :=
+begin
+  ext,
+  apply completion.induction_on x,
+  { apply is_closed_eq,
+    continuity },
+  { simp [normed_group_hom.mem_ker, completion.coe_zero] }
+end
+
+lemma normed_group.dense_range_to_compl : dense_range (to_compl : G → completion G) :=
+completion.dense_inducing_coe.dense
+
+@[simp]
+lemma normed_group_hom.completion_to_compl (f : normed_group_hom G H) :
+  f.completion.comp to_compl = to_compl.comp f :=
+begin
+  ext x,
+  change f.completion x = _,
+  simpa
+end
+
+lemma normed_group_hom.norm_completion_le (f : normed_group_hom G H) : ∥f.completion∥ ≤ ∥f∥ :=
```

Comment:
For normed groups this is an equality, right?

Reply:
Even semi-normed groups actually, but we were too quick in LTE (and some infrastructure was missing). Thanks for pointing this out.